### ISSUE TITLE
Type annotation ignored when type is partially unknown

### ISSUE THREAD

**Issue by parched** [microsoft/pyright#4803]
While upgrading pyright I noticed this new error which I believe should not be an error.

**Screenshots or Code**
```python

def a_test(x: int | str) -> str:
    y: str = x.upper()  # type: ignore
    return y
```
![image](https://user-images.githubusercontent.com/5975405/226158420-7933cb75-13c6-4dbf-88a0-d01a088e473c.png)

**VS Code extension or command-line**
commandline it started happening in 1.1.281

**Comment by erictraut** [comment]
The behavior is as designed. The value of `x.upper()` is unknown when `x` is an `int`. 

Ignoring the type violation with a `# type: ignore` does not change the result of the violation; it simply suppresses the error. If you know that `x` is not an int on that code path, you should insert an `assert` statement.

```python
def a_test(x: int | str) -> str:
    assert(isinstance(x, str))
    y: str = x.upper()
    return y
```

**Comment by parched** [comment]
Why does this typecheck fine then?
```
def a_test(x: int | str) -> str:
    y: str = x  # type: ignore
    return y
```

My actual code looks more like this
```
def _to_json_primitive(result: np.ndarray | np.void) -> _Json:
    names: tuple[str, ...] | None = result.dtype.names  # type: ignore
    if names is not None:
        if names[-1] == DISCRIMINANT_NAME:
            ...
```
So I can't use an assert because it's generic, but it looks like I can use `cast` *and* `# type: ignore` to get around it.

**Comment by erictraut** [comment]
> Why does this typecheck fine then?

It doesn't type check. There is an error, and you're suppressing it with a `# type: ignore` comment.

I presume that you're also using strict type checking here. If you disable the `reportUnknownVariableType` check and leave the `# type: ignore` in place, you won't see any error.

```python
# pyright: strict, reportUnknownVariableType=false

def a_test(x: int | str) -> str:
    y: str = x.upper()  # type: ignore
    return y
```

**Comment by parched** [comment]
> It doesn't type check. There is an error, and you're suppressing it with a `# type: ignore` comment.

Apologies, let me rephrase. Why does `# type: ignore` work in this case but not the original?

I can understand there may be reasons from a developer's perspective, I just can't understand from a user's perspective.

**Comment by erictraut** [comment]
The `# type: ignore` comment simply suppresses all errors on that line. It doesn't change any type evaluation behaviors.

The `reportUnknownVariableType` is reported on the `return` statement, which is not on the same line as the `# type: ignore`, so it is not suppressed. By adding `# pyright: reportUnknownVariableType=false`, you will suppress all `reportUnknownVariableType` errors in the file.

You can choose to add a second `# type: ignore` if you want to suppress the error on the next line as well.

```python
# pyright: strict

def a_test(x: int | str) -> str:
    y: str = x.upper()  # type: ignore
    return y # type: ignore
```

FWIW, I _never_ use `# type: ignore` in my code. I find that there is always a better (less fragile) solution that suppressing errors.

**Comment by parched** [comment]
The problem is that I use `y` multiple times and don't want to suppress every use site. I just want to say to pyright "Ignore what the incomplete numpy type stubs are telling you, **this** is the actual type of the variable, just believe me, OK?"

Agree, I prefer not to use `# type: ignore` and being more specific with an `assert` if possible or `# pyright: ignore [reportXXXX]`.
I find, however, that
* copilot always suggests the wrong, or non-existent, diagnostic
* it takes me multiple tries to get the spelling correct myself (how many "c"s and "s"s in necessary again? 😄),
* then something long like `# pyright: ignore [reportUnknownArgumentType]` has caused the line to overflow, so the formatter puts it on the next line, now it no longer applies.

**Comment by erictraut** [comment]
If you're using incomplete type stubs, then it's going to be difficult to use strict type checking because the stubs will result in `Unknown` types.

**Comment by vnmabus** [comment]
I am seeing the same behavior reported here. If I have:

```python
a: int
a = ""  # type: ignore
```

then the type of `a` is still reported as `int` afterwards (not `int | str`), which is as intended.

However, if one has instead:
```python
a: int
a = "".unknown_method()  # type: ignore
```

then the type of `a` is reported as `Unknown | int` afterwards!!

This is problematic if you have dynamic methods but you know the resulting type of the variable, as now you have to use casts and/or asserts after every method call just to please Pyright, when it shouldn't have unilaterally **changed** the declared type in the first place (specially if it **doesn't** do it in other cases!).

**Comment by vnmabus** [comment]
@erictraut Sorry for pinging you about this (I know you are busy), but I would really want to know if this difference of behavior (`Unknown` modifying the declared type, while it does not happen with other types) is a bug, or if there is any reason for that. Should I maybe open a new issue for this?
